src/Data/Array/Mutable/Unlifted/Linear.hs

```diff
@@ -158,6 +163,16 @@ toList = unArray# Prelude.$ \arr ->
         case GHC.runRW# (GHC.readArray# arr i#) of
           (# _, ret #) -> ret : go (i Prelude.+ 1) len arr
 
+-- | /O(1)/ Convert an 'Array#' to an immutable 'Vector' (from 'vector' package).
+freeze :: forall a. Array# a #-> Ur (Vector.Vector a)
```

Comment:
I think it would have been natural, here, to convert to an unlifted immutable array. And post-convert to `Vector` in the unlifted case instead.

See [`unsafeFreezeArray#`](https://hackage.haskell.org/package/base-4.12.0.0/docs/GHC-Exts.html#v:unsafeFreezeArray-35-)

Reply:
MVector's constructor is as follows:

```haskell
data MVector s a = MVector {-# UNPACK #-} !Int
                           {-# UNPACK #-} !Int
                           {-# UNPACK #-} !(MutableArray s a)
        deriving ( Typeable )
```

That `MutableArray` is from the `primitive` package:

```haskell
-- | Mutable boxed arrays associated with a primitive state token.
data MutableArray s a = MutableArray
  { marray# :: MutableArray# s a }
  deriving ( Typeable )
```

So, that forces us to go through `MutableArray`. The `Vector` constructor work as you mentioned:

```
data Vector a = Vector {-# UNPACK #-} !Int
                       {-# UNPACK #-} !Int
                       {-# UNPACK #-} !(Array a)
        deriving ( Typeable )
```

And it'd be ideal to use it, however that constructor is not exported.

Reply:
I now understand why you are asking for this, and I agree. Freezing an unlifted mutable array to an unlifted immutable array makes sense. I did the change.

Which caused unlifted arrays freeze function to have a different signature:

```haskell
freeze :: (GHC.Array# a -> b) -> Array# a #-> Ur b
```

Because I couldn't put an unlifted type inside a `Ur`. We are using the same pattern on `unArray#` function too, so I think it is fine.

Reply:
Yes, it's safe. `Ur a` is isomorphic to `forall k. (a -> k) #-> k` (exercise!).

Ultimately, we probably want to expose an unlifted unrestricted thing. There is no obstacle except time and syntax.

Reply:
Here is the corresponding issue, by the way: https://gitlab.haskell.org/ghc/ghc/-/issues/18490 .